internal/controller/numaflowcontroller/numaflowcontroller_controller.go

```diff
@@ -252,17 +252,46 @@ func (r *NumaflowControllerReconciler) reconcile(
 		controllerutil.AddFinalizer(controller, common.FinalizerName)
 	}
 
-	_, deploymentExists, err := r.getNumaflowControllerDeployment(ctx, controller)
+	existingDeployment, deploymentExists, err := r.getNumaflowControllerDeployment(ctx, controller)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
+	newVersion := controller.Spec.Version
+	targetObjs, err := determineTargetObjects(controller, newVersion)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to determine the target objects for the new version %s: %w", newVersion, err)
+	}
+
+	numaLogger.Debugf("found %d target objects associated with NumaflowController version %s", len(targetObjs), newVersion)
+
+	// If a Deployment already exists for the NumaflowController and the new version differs from the current Deployment version,
+	// delete all the NumaflowController child resources and recreate them at the next reconciliation
+	if deploymentExists {
+		currentVersion, err := getControllerDeploymentVersion(existingDeployment)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error while getting the version of the existing NumaflowController Deployment: %w", err)
+		}
+
+		childResourcesNeedToBeDeleted := controller.Spec.Version != currentVersion
+		if childResourcesNeedToBeDeleted {
+			numaLogger.Debugf("current Deployment image tag=%q differs from desired %q", currentVersion, controller.Spec.Version)
+
+			err := r.deleteNumaflowControllerChildren(controller, currentVersion, namespace)
+			if err != nil {
+				return ctrl.Result{}, fmt.Errorf("error deleting NumaflowController child resources: %w", err)
+			}
+
+			return ctrl.Result{}, nil
```

Comment:
oh, did you decide not to explicitly requeue because you are figuring that we'll automatically re-reconcile due to the change in child resources that we're watching?

Reply:
That's correct, I didn't think requeuing was necessary unless you can think of a case in which reconciliation would not re-execute. If we want to be 100% sure, I can explicitely requeue.

Reply:
On a second thought, would it be better to explicitly requeue to start the next reconciliation faster?

Reply:
I think we should be okay as is. Can we just add a comment at that return statement about the assumption we're making?

Worst case, Chandan's new [PR ](https://github.com/numaproj/numaplane/pull/460) will ensure we re-reconcile every 15 minutes.